server/src/main/java/io/deephaven/server/notebook/FilesystemStorageServiceGrpcImpl.java

```diff
@@ -0,0 +1,188 @@
+        Path resolved = root.resolve(relativePath).normalize();
+        if (resolved.startsWith(root)) {
+            return Optional.of(resolved);
+        }
+        return Optional.empty();
+    }
+    private Path resolveOrThrow(String relativePath) {
+        return resolve(relativePath).orElseThrow(() -> GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Invalid path: " + relativePath));
+    }
+    @Override
+    public void listItems(ListItemsRequest request, StreamObserver<ListItemsResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            ListItemsResponse.Builder builder = ListItemsResponse.newBuilder();
+            PathMatcher matcher = request.hasFilterGlob() ? makeMatcher(request.getFilterGlob()) : ignore -> true;
+            Path dir = resolveOrThrow(request.getPath());
+            try (Stream<Path> list = Files.list(dir)) {
+                for (Path p : (Iterable<Path>) list::iterator) {
+                    if (!matcher.matches(dir.relativize(p))) {
+                        continue;
+                    }
+                    builder.addItems(FileInfo.newBuilder()
+                            .setPath(p.getFileName().toString())
+                            .setSize(Files.isDirectory(p) ? 0 : Files.size(p))
+                            .setKind(Files.isDirectory(p) ? FileKind.DIRECTORY : FileKind.FILE)
+                            .build());
+                }
+            } catch (NoSuchFileException noSuchFileException) {
+                throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION, "Directory does not exist");
+            }
+            responseObserver.onNext(builder.build());
+            responseObserver.onCompleted();
+        });
+    }
+
+    private PathMatcher makeMatcher(String filterGlob) {
+        if (filterGlob.contains("**")) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Bad glob, only single `*`s are supported");
+        }
+        if (filterGlob.contains("/")) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Bad glob, only the same directory can be checked");
+        }
+        try {
+            return FileSystems.getDefault().getPathMatcher("glob:" + filterGlob);
+        } catch (PatternSyntaxException e) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT, "Bad glob, can't parse expression: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void fetchFile(FetchFileRequest request, StreamObserver<FetchFileResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final byte[] bytes;
+            try {
+                bytes = Files.readAllBytes(resolveOrThrow(request.getPath()));
+            } catch (NoSuchFileException noSuchFileException) {
+                throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION, "File does not exist");
+            }
```

Comment:
Should we treat this as a non-export to prevent too many concurrent callers? The same comment applies to all of the gRPCs that interact with the file system.

Reply:
Not following - are you referring to gRPC's guarantees of a single stream always having messages handled serially? If so, these are all unary calls, so should be safe to handle concurrently.